src/Workspaces/Core/MSBuild/Rpc/RpcClient.cs

```diff
@@ -0,0 +1,150 @@
+                    {
+                        completionSource.SetResult(null);
+                    }
+                    else
+                    {
+                        try
+                        {
+                            // response.Value might be null if the response was in fact null.
+                            var result = response.Value?.ToObject(expectedType);
+                            completionSource.SetResult(result);
+                        }
+                        catch (Exception ex)
+                        {
+                            completionSource.SetException(new Exception("Unable to deserialize the result", ex));
+                        }
+                    }
+                }
+            }
+
+            // We've disconnected, so cancel any remaining outstanding tasks
+            foreach (var (request, _) in _outstandingRequests.Values)
+                request.SetException(new System.Exception("The server disconnected unexpectedly."));
+
+            Disconnected?.Invoke(this, EventArgs.Empty);
+        });
+    }
+
+    public void Shutdown()
+    {
+        _shutdownTokenSource.Cancel();
+    }
+
+    public Task InvokeAsync(int targetObject, string methodName, List<object?> parameters, CancellationToken cancellationToken)
+    {
+        // We'll ignore the return value of InvokeCoreAsync in this case, since it'll just be filled in with null
+        return InvokeCoreAsync(targetObject, methodName, parameters, expectedReturnType: null, cancellationToken);
+    }
+
+    public async Task<T?> InvokeNullableAsync<T>(int targetObject, string methodName, List<object?> parameters, CancellationToken cancellationToken) where T : class
+    {
+        var result = await InvokeCoreAsync(targetObject, methodName, parameters, expectedReturnType: typeof(T), cancellationToken).ConfigureAwait(false);
+        return (T?)result;
+    }
+
+    public async Task<T> InvokeAsync<T>(int targetObject, string methodName, List<object?> parameters, CancellationToken cancellationToken) where T : notnull
+    {
+        var result = await InvokeCoreAsync(targetObject, methodName, parameters, expectedReturnType: typeof(T), cancellationToken).ConfigureAwait(false);
+        Contract.ThrowIfNull(result, "We expected a non-null result but got null back.");
+        return (T)result;
+    }
+
+    private async Task<object?> InvokeCoreAsync(int targetObject, string methodName, List<object?> parameters, Type? expectedReturnType, CancellationToken cancellationToken)
+    {
+        int requestId;
+        var requestCompletionSource = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        // Loop until we successfully have a new index for this; practically we don't expect this to ever collide, since that'd mean we'd have
+        // billions of requests outstanding, but...
+        do
+        {
+            requestId = Interlocked.Increment(ref _nextRequestId);
```

Comment:
Seems like the first requestId here will be 1, whereas the target object id start at 0.  Does it matter?  Probably not but slightly strange we're doing them differently?

Reply:
I think they both started at one, and then when hardcoding the target index starting at zero seemed superficially better. I can change to match (but that's the only reason) or have the target index start at 1 if you'd like.